src/cocotb/__init__.py

```diff
@@ -136,7 +135,7 @@ def _task_done_callback(task: "cocotb.task.Task[Any]") -> None:
     if e is None:
         return
     # there was a failure and no one is watching, fail test
-    elif isinstance(e, (TestSuccess, AssertionError)):
+    elif isinstance(e, (_TestSuccess, _Failed, AssertionError)):
```

Comment:
This was a bug since before cocotb 2.0. We weren't handling `pytest.raises` and `pytest.warns` errors in Tasks. Whoops...